### PR TITLE
add bounded concurrency for tag lookup and untag

### PR DESCRIPTION
**Pull request by microyahoo**
add concurrency limits for tag lookup and untag
    
Harbor is using the distribution for it's (harbor-registry) registry component.
    
The harbor GC will call into the registry to delete the manifest, which in turn then does a lookup for all tags that reference the deleted manifest.
To find the tag references, the registry will iterate every tag in the repository and read it's link file to check if it matches the deleted manifest (i.e. to see if uses the same sha256 digest). So, the more tags in repository, the worse the performance will be (as there will be more s3 API calls occurring for the tag directory lookups and tag file reads).
    
Therefore, we can use concurrent lookup and untag to optimize performance as described in https://github.com/goharbor/harbor/issues/12948.
    
P.S. This optimization was originally contributed by @Antiarchitect, now I would like to take it over.
Thanks @Antiarchitect's efforts with PR https://github.com/distribution/distribution/pull/3890.
    
Signed-off-by: Liang Zheng <zhengliang0901@gmail.com>

### CODE REVIEW

File: registry/handlers/manifests.go
Comment (milosgajdos) [inline registry/handlers/manifests.go:498]: We need 2 things here (given that the `errgroup` goroutines return `nil`)
1. Add a comment explaining why we're not checking `error`
2. Add a marked for the linter to ignore the `errcheck` -> `// nolint:errcheck

File: registry/handlers/manifests.go
Comment (milosgajdos) [inline registry/handlers/manifests.go:493]: Also, we need a `mutex` here; we can't be appending `imh.Errors` concurrently without a mutex

**Comment by milosgajdos** [comment]
@corhere do you mind batting an eye on this? It looks a ok to me, but I need to give it another look.

**Comment by milosgajdos** [comment]
@Jamstah @squizzi you folks mind having a look at this?

File: registry/storage/registry.go
Comment (squizzi) [inline registry/storage/registry.go:14]: How did you arrive at 8 for the default?

File: registry/handlers/app.go
Comment (corhere) [inline registry/handlers/app.go:189]: Every new contributor tries to add `snake_case` config fields for some reason. You can't do that.
https://github.com/distribution/distribution/blob/df983747649396a7377a4b428ad3a1aebf7a9001/configuration/configuration.go#L16-L17
```suggestion
		l, ok := p["concurrencylimit"]
```

File: registry/handlers/context.go
Comment (corhere) [inline registry/handlers/context.go:38]: Don't extend this struct. It runs contrary to current best practices against contexts, and adding new fields makes it more difficult to refactor away in the future. Also, there is no need for a mutex here as a given instance of the struct is never shared among multiple concurrent requests.

File: registry/handlers/manifests.go
Comment (corhere) [inline registry/handlers/manifests.go:501]: Try to avoid using nolint annotations when more idiomatic constructs are available.
```suggestion
	// imh already records all errors, so ignore the error of Wait()
	_ = g.Wait()
```

File: registry/handlers/manifests.go
Comment (corhere) [reply registry/handlers/manifests.go:494]: The mutex doesn't have to be on `imh` (read: I will not approve this PR unless it is changed.)
```go
var (
	errs []error
	mu   sync.Mutex
)
for _, tag := range referencedTags {
	g.Go(func() error {
		if err := tagService.Untag(imh, tag); err != nil {
			mu.Lock()
			errs = append(errs, err)
			mu.Unlock()
		}
		return nil
	})
}
_ = g.Wait()
imh.Errors = errs
```

File: registry/handlers/app.go
Comment (corhere) [inline registry/handlers/app.go:193]: What are the semantics of an explicit concurrency limit of `0`? `-1`?

File: registry/storage/tagstore.go
Comment (corhere) [inline registry/storage/tagstore.go:179]: Why are you using a mutex _scoped to the whole tagStore_ to synchronize access to a _local_ variable?

**Review by corhere — CHANGES_REQUESTED** [review]

File: registry/handlers/app.go
Comment (microyahoo) [reply registry/handlers/app.go:193]: During the initialization of the tagStore, a judgment is made, and if it is greater than 0, it will be set; otherwise, the concurrency limit is set to the default value.
```golang
limit := DefaultConcurrencyLimit
if repo.tagLookupConcurrencyLimit > 0 {
	limit = repo.tagLookupConcurrencyLimit
}
```

Comment (milosgajdos) [reply registry/handlers/app.go:193]: We should document that in the docs FWIW

Comment (microyahoo) [reply registry/handlers/app.go:193]: > We should document that in the docs FWIW

done

Comment (corhere) [reply registry/handlers/app.go:197]: Defining all non-positive integers to mean `GOMAXPROCS` does not leave any values to signify the other possible special case: unbounded concurrency / no limit. To be clear, **I don't think an option for unbounded concurrency should be added as part of this PR.** I simply want to reserve the negative numbers for future use by having it be an error to set `concurrencylimit` to a negative value.

File: registry/storage/tagstore.go
Comment (corhere) [inline registry/storage/tagstore.go:152]: If any tag lookups fail the whole result is discarded, so any work performed after the first failure is a waste of time and resources. Cancelation is trivial with errgroup when the work accepts a context, so we may as well implement it.
```suggestion
	g, ctx := errgroup.WithContext(ctx)
```

Also, break the loop early.
```go
for _, tag := range allTags {
	if ctx.Err() != nil {
		break
	}
	...
}
```

File: registry/handlers/app.go
Comment (corhere) [reply registry/handlers/app.go:197]: I suggest that we _do not handle_ unbounded concurrency. At least not as part of this PR. What I am suggesting is the following:

| `concurrencylimit` value | Result |
|--------|--------|
| (unset) | Tag lookups are limited to `GOMAXPROCS` concurrency (the default) |
| `0` | Tag lookups are limited to `GOMAXPROCS` concurrency (the default) |
| Positive integer, e.g. `3` | Tag lookups are limited to the value, e.g. `3` | 
| Negative integer, e.g. `-1` | Configuration error |

Making negative numbers an error today guarantees that no valid registry configuration can have a `concurrencylimit` option set to a negative number. That gives us the freedom to redefine negative values to mean something valid in the future _as a non-breaking change_. One potential _future_ definition of negative `concurrencylimit` values is to configure unbounded concurrency. But we don't need that today.

Note that setting `concurrencylimit: 0` should be allowed. It is very useful to have an "explicit default" value for distribution configuration as it allows the user to be explicit about their intent to use `GOMAXPROCS` as the concurrency limit in their configuration, or to override e.g. `concurrencylimit: 4` in the YAML config by setting the environment variable `REGISTRY_TAG_CONCURRENCYLIMIT=0`.

Comment (microyahoo) [reply registry/handlers/app.go:197]: Thanks for your nice explanation, I have updated it, PTAL.

**Review by corhere — APPROVED** [review]
LGTM, thanks!

**Review by milosgajdos — APPROVED** [review]
Thanks for a nice contribution @microyahoo

**Comment by Antiarchitect** [comment]
@microyahoo Congrats! You made it!

**Comment by dkulchinsky** [comment]
Congrats @microyahoo! very excited to see this getting merged and thanks to @Antiarchitect for the initial work.

**Pull request merged**
